File: TheepotAndroid/app/src/main/kotlin/nl/bsodetheepot/mobile/data/models/Prikbord.kt
package nl.bsodetheepot.mobile.data.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import nl.bsodetheepot.mobile.data.services.DateUtils
import java.time.Instant

@Serializable
enum class PrikbordPrioriteit {
    @SerialName("normaal") NORMAAL,
    @SerialName("belangrijk") BELANGRIJK,
    @SerialName("urgent") URGENT;

    val label: String
        get() = when (this) {
            NORMAAL -> "Normaal"
            BELANGRIJK -> "Belangrijk"
            URGENT -> "Urgent"
        }
}

@Serializable
data class NaamOnly(val naam: String)

/** Tabel `prikbord_berichten`. */
@Serializable
data class PrikbordBericht(
    val id: String,
    @SerialName("locatie_naam") val locatieNaam: String,
    val titel: String,
    val inhoud: String,
    val prioriteit: PrikbordPrioriteit,
    @SerialName("aangemaakt_door") val aangemaaktDoor: String? = null,
    @SerialName("aangemaakt_op") val aangemaaktOp: String,
    val verloopdatum: String? = null,
    @SerialName("gelezen_door") val gelezenDoor: List<String>? = null,
    val profielen: NaamOnly? = null,
) {
    val auteurNaam: String get() = profielen?.naam ?: "Onbekend"

    val isVerlopen: Boolean
        get() {
            val v = verloopdatum ?: return false
            val datum = runCatching { Instant.parse(v) }.getOrNull()
                ?: DateUtils.parseDateStr(v)?.atStartOfDay(DateUtils.amsterdam)?.toInstant()
                ?: return false
            return datum.isBefore(Instant.now())
        }
}

File: TheepotMobile/TheepotMobile/Models/Prikbord.swift
import Foundation

enum PrikbordPrioriteit: String, Codable, CaseIterable {
    case normaal, belangrijk, urgent

    var label: String {
        switch self {
        case .normaal: return "Normaal"
        case .belangrijk: return "Belangrijk"
        case .urgent: return "Urgent"
        }
    }
}

/// Tabel `prikbord_berichten`.
struct PrikbordBericht: Codable, Identifiable {
    let id: String
    var locatieNaam: String
    var titel: String
    var inhoud: String
    var prioriteit: PrikbordPrioriteit
    var aangemaaktDoor: String?
    let aangemaaktOp: String
    var verloopdatum: String?
    var gelezenDoor: [String]?
    let profielen: NaamOnly?

    enum CodingKeys: String, CodingKey {
        case id
        case locatieNaam = "locatie_naam"
        case titel, inhoud, prioriteit
        case aangemaaktDoor = "aangemaakt_door"
        case aangemaaktOp = "aangemaakt_op"
        case verloopdatum
        case gelezenDoor = "gelezen_door"
        case profielen
    }

    var auteurNaam: String { profielen?.naam ?? "Onbekend" }

    var isVerlopen: Bool {
        guard let verloopdatum, let datum = ISO8601DateFormatter.theepot.date(from: verloopdatum) ?? DateFormatter.theepotDate.date(from: verloopdatum) else {
            return false
        }
        return datum < Date()
    }
}

struct NaamOnly: Codable {
    let naam: String
}

extension ISO8601DateFormatter {
    static let theepot = ISO8601DateFormatter()
}

extension DateFormatter {
    static let theepotDate: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.timeZone = TimeZone(identifier: "Europe/Amsterdam")
        return f
    }()
}
